js/server/types/User.js

```diff
@@ -0,0 +1,228 @@
+const _ = require('lodash');
+const moment = require('moment');
+const { UserTicket } = require('./UserTicket');
+const config = require('../config');
+
+// currently just a single source of truth for the User data structure.
+// May be useful for migrating the user mutation methods to in the future.
+class User {
+  constructor () {
+    this.tickets = [];
+    this.claimableRewardsOnWithdrawnAssets = 0;
+    this.dispensed = 0;
+    this.forfeited = 0;
+    this.currentTotalClaimableReward = 0;
+    this.reservedReward = 0;
+    this.totalTicketsAmountSum = 0;
+    this.claimableCommissionsOnDelegatorRewards = 0;
+    this.totalRewardAtMaturity = 0;
+    this.ticketAmountAtMaturity = 0;
+    this.yieldAtMaturity = 0;
+    this.nextRewardShare = 0;
+    this.currentYieldOnTickets = 0;
+    // dates
+    this.maturityDate = undefined;
+    this.maturityDateISO = undefined;
+    this.yearsToMaturity = undefined;
+    this.currentAPYOnTickets = undefined;
+    this.maturityDateMs = undefined;
+    // timevalues
+    this.futureReward = undefined;
+    this.nextReward = undefined;
+    this.nextRewardProjectedFutureReward = undefined;
+    this.nextRewardProjectedAPYOnTickets = undefined;
+  }
+
+  updateMaturityTimeProps (userAtFinalTimestamp, currentTimestampInMs) {
+    this.maturityDate = userAtFinalTimestamp.maturityDate;
+    this.maturityDateISO = userAtFinalTimestamp.maturityDateISO;
+    const msToMaturity =
+      userAtFinalTimestamp.maturityDateMs - currentTimestampInMs;
+    this.yearsToMaturity = msToMaturity / 1000 / 60 / 60 / 24 / 365;
+    this.currentAPYOnTickets =
+      this.currentYieldOnTickets / this.yearsToMaturity;
+  }
+
+  cloneWith (props) {
+    let next = new User();
+    Object.assign(next, this);
+    Object.assign(next, props);
+    return next;
+  }
+
+  refreshTotalTicketAmountSum () {
+    this.totalTicketsAmountSum = _.sum(this.tickets.map(t => t.amount));
+  }
+
+  updateRewards (timestampTicketsAmountSum) {
+    this.refreshTotalTicketAmountSum();
+    this.currentTotalClaimableReward =
+      this.claimableRewardsOnWithdrawnAssets +
+      _.sum(this.tickets.map(t => t.reward * t.mul));
+    this.reservedReward =
+      this.claimableRewardsOnWithdrawnAssets +
+      _.sum(this.tickets.map(t => t.reward));
+    this.nextRewardShare =
+      this.totalTicketsAmountSum / timestampTicketsAmountSum;
+  }
+
+  updateUserMaturityDates (
```

Comment:
Would prefer these were all named args, when it gets to 4 args it gets confusing from the outside.

Reply:
Agreed. Optimally, they would be defined as typescript types as well